lib/Runtime/Language/SourceTextModuleRecord.cpp

```diff
@@ -165,25 +217,245 @@ namespace Js
+                    // We are not rejecting ambiguous resolution at this time.
+                    if (starExportedNames != nullptr)
+                    {
+                        starExportedNames->Map([&](PropertyId propertyId) {
+                            if (propertyId != defaultPropertyId && !localNames->Has(propertyId))
+                            {
+                                localNames->Prepend(propertyId);
+                            }
+                        });
+                    }
+                }
+#if DBG
+                else
+                {
+                    AssertMsg(false, "dependent modules should have been initialized");
+                }
+#endif
+            });
+        }
+        exportedNames = localNames;
+        return localNames;
+    }
+
+    // return false when "ambiguous". 
+    // otherwise nullptr means "null" where we have circular reference/cannot resolve.
+    bool SourceTextModuleRecord::ResolveExport(PropertyId exportName, ResolveSet* resolveSet, ExportModuleRecordList* exportStarSet, ModuleNameRecord** exportRecord)
+    {
+        ArenaAllocator* allocator = EnsureTempAllocator();
+        if (resolvedExportMap == nullptr)
         {
-            resolveSet = (ResolutionDictionary*)AllocatorNew(ArenaAllocator, allocator, ResolutionDictionary, allocator);
+            resolvedExportMap = AllocatorNew(ArenaAllocator, allocator, ResolvedExportMap, allocator);
         }
+        if (resolvedExportMap->TryGetReference(exportName, exportRecord))
+        {
+            return true;
+        }
+        // TODO: use per-call/loop allocator?
         if (exportStarSet == nullptr)
         {
-            exportStarSet = (ResolveSet*)AllocatorNew(ArenaAllocator, allocator, ResolveSet, allocator);
+            exportStarSet = (ExportModuleRecordList*)AllocatorNew(ArenaAllocator, allocator, ExportModuleRecordList, allocator);
+        }
+        if (resolveSet == nullptr)
+        {
+            resolveSet = (ResolveSet*)AllocatorNew(ArenaAllocator, allocator, ResolveSet, allocator);
+        }
+
+        *exportRecord = nullptr;
+        bool hasCircularRef = false;
+        resolveSet->MapUntil([&](ModuleNameRecord moduleNameRecord) {
+            if (moduleNameRecord.module == this && moduleNameRecord.bindingName == exportName)
+            {
+                *exportRecord = nullptr;
+                hasCircularRef = true;
+                return true;
+            }
+            return false;
+        });
+        if (hasCircularRef)
+        {
+            Assert(*exportRecord == nullptr);
+            return true;
         }
-        Assert(false);
-        return false;
+        resolveSet->Prepend({ this, exportName });
+
+        if (localExportRecordList != nullptr)
+        {
+            PropertyId localNameId = Js::Constants::NoProperty;
+            localExportRecordList->MapUntil([&](ModuleExportEntry exportEntry) {
```

Comment:
Fair enough. LGTM.

--Paul

From: Yong Qu [mailto:notifications@github.com]
Sent: Friday, March 4, 2016 3:57 PM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Paul Leathers <pleath@microsoft.com>
Subject: Re: [ChakraCore] module runtime change #3 (#432)


In lib/Runtime/Language/SourceTextModuleRecord.cpp<https://github.com/Microsoft/ChakraCore/pull/432#discussion_r55106661>:

>          }

> -        Assert(false);

> -        return false;

> +        resolveSet->Prepend({ this, exportName });

> +

> +        if (localExportRecordList != nullptr)

> +        {

> +            PropertyId localNameId = Js::Constants::NoProperty;

> +            localExportRecordList->MapUntil([&](ModuleExportEntry exportEntry) {

No but it might be better to come in from a later change.

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/432/files#r55106661>.